gnovm/pkg/doc/json_doc.go

```diff
@@ -0,0 +1,246 @@
+			Const:     true,
+			Values:    d.extractValueSpecs(pkg, value.Decl.Specs),
+			Doc:       string(pkg.Markdown(value.Doc)),
+		})
+	}
+
+	for _, value := range pkg.Vars {
+		jsonDoc.Values = append(jsonDoc.Values, &JSONValueDecl{
+			Signature: mustFormatNode(d.pkgData.fset, value.Decl),
+			Const:     false,
+			Values:    d.extractValueSpecs(pkg, value.Decl.Specs),
+			Doc:       string(pkg.Markdown(value.Doc)),
+		})
+	}
+
+	for _, fun := range pkg.Funcs {
+		jsonDoc.Funcs = append(jsonDoc.Funcs, &JSONFunc{
+			Name:      fun.Name,
+			Signature: mustFormatNode(d.pkgData.fset, fun.Decl),
+			Doc:       string(pkg.Markdown(fun.Doc)),
+			Params:    d.extractFuncParams(fun),
+			Results:   d.extractFuncResults(fun),
+		})
+	}
+
+	for _, typ := range pkg.Types {
+		jsonDoc.Types = append(jsonDoc.Types, &JSONType{
+			Name:      typ.Name,
+			Signature: mustFormatNode(d.pkgData.fset, typ.Decl),
+			Doc:       string(pkg.Markdown(typ.Doc)),
+		})
+
+		// constructors for this type
+		for _, fun := range typ.Funcs {
+			jsonDoc.Funcs = append(jsonDoc.Funcs, &JSONFunc{
+				Name:      fun.Name,
+				Signature: mustFormatNode(d.pkgData.fset, fun.Decl),
+				Doc:       string(pkg.Markdown(fun.Doc)),
+				Params:    d.extractFuncParams(fun),
+				Results:   d.extractFuncResults(fun),
+			})
+		}
+
+		for _, meth := range typ.Methods {
+			jsonDoc.Funcs = append(jsonDoc.Funcs, &JSONFunc{
+				Type:      typ.Name,
+				Name:      meth.Name,
+				Signature: mustFormatNode(d.pkgData.fset, meth.Decl),
+				Doc:       string(pkg.Markdown(meth.Doc)),
+				Params:    d.extractFuncParams(meth),
+				Results:   d.extractFuncResults(meth),
+			})
+		}
+	}
+
+	return jsonDoc, nil
+}
+
+func (d *Documentable) extractFuncParams(fun *doc.Func) []*JSONField {
+	params := []*JSONField{}
+	for _, param := range fun.Decl.Type.Params.List {
+		// parameters can be of the format: (a, b int, c string)
+		// so we need to iterate over the names
+		for _, name := range param.Names {
+			field := &JSONField{
+				Name: name.Name,
+				Type: mustFormatNode(d.pkgData.fset, param.Type),
+			}
+
+			params = append(params, field)
+		}
+	}
+
+	return params
+}
+
+func (d *Documentable) extractFuncResults(fun *doc.Func) []*JSONField {
+	results := []*JSONField{}
+	if fun.Decl.Type.Results != nil {
+		for _, result := range fun.Decl.Type.Results.List {
+			if len(result.Names) == 0 {
+				// if there are no names, then the result is an unnamed return
+				result := &JSONField{
+					Name: "",
+					Type: mustFormatNode(d.pkgData.fset, result.Type),
+				}
+				results = append(results, result)
+			} else {
+				// results can be of the format: (a, b int, c string)
+				// so we need to iterate over the names
+				for _, name := range result.Names {
+					result := &JSONField{
+						Name: name.Name,
+						Type: mustFormatNode(d.pkgData.fset, result.Type),
+					}
+					results = append(results, result)
+				}
+			}
+		}
+	}
+	return results
+}
```

Comment:
I think these can and should be merged into one, also because like the return values can have no names, so can the parameters: `func(int, bool) string` is a valid function signature in go

Reply:
Good point. Done. https://github.com/gnolang/gno/pull/3459/commits/cd2960918df8328e91ae5c1900856696cceb692b